client/state-db/src/lib.rs

```diff
@@ -339,30 +342,39 @@ impl<BlockHash: Hash + MallocSizeOf, Key: Hash + MallocSizeOf> StateDbSync<Block
 		if let Some(ref mut pruning) = self.pruning {
 			pruning.note_canonical(hash, &mut commit);
 		}
-		self.prune(&mut commit);
+		self.prune(&mut commit)?;
 		Ok(commit)
 	}
 
 	fn best_canonical(&self) -> Option<u64> {
 		self.non_canonical.last_canonicalized_block_number()
 	}
 
-	fn is_pruned(&self, hash: &BlockHash, number: u64) -> bool {
-		match self.mode {
+	fn is_pruned(&self, hash: &BlockHash, number: u64) -> IsPruned {
+		let is_pruned = match self.mode {
 			PruningMode::ArchiveAll => false,
 			PruningMode::ArchiveCanonical | PruningMode::Constrained(_) => {
 				if self.best_canonical().map(|c| number > c).unwrap_or(true) {
 					!self.non_canonical.have_block(hash)
 				} else {
-					self.pruning.as_ref().map_or(false, |pruning| {
-						number < pruning.pending() || !pruning.have_block(hash)
-					})
+					match self.pruning.as_ref() {
+						None => false,
```

Comment:
This should result in a` Maybe`

Reply:
This is also subject to the original logic, the rationale behind it may be `self.pruning` is only set to `Some` for the `PruningMode::Constrained` mode, and both `PruningMode::ArchiveAll` and `PruningMode::ArchiveCanonical` won't prune the state of canonicalized block